core/executor/src/allocator.rs

```diff
@@ -78,35 +115,46 @@ impl FreeingBumpHeapAllocator {
 	}
 
 	/// Gets requested number of bytes to allocate and returns a pointer.
+	/// The minimum size which can be allocated is 8 bytes.
 	/// The maximum size which can be allocated at once is 16 MiB.
 	pub fn allocate(&mut self, size: u32) -> Result<u32> {
 		if size > MAX_POSSIBLE_ALLOCATION {
 			return Err(Error::RequestedAllocationTooLarge);
 		}
 
-		let size = size.max(8);
+		let size = size.max(MIN_POSSIBLE_ALLOCATION);
 		let item_size = size.next_power_of_two();
-		if item_size + 8 + self.total_size > self.max_heap_size {
+		if item_size + PREFIX_SIZE + self.total_size > self.max_heap_size {
 			return Err(Error::AllocatorOutOfSpace);
 		}
 
 		let list_index = (item_size.trailing_zeros() - 3) as usize;
 		let ptr: u32 = if self.heads[list_index] != 0 {
 			// Something from the free list
 			let item = self.heads[list_index];
+			let ptr = item + PREFIX_SIZE;
+			assert!(
+				ptr + item_size <= self.max_heap_size,
+				"Pointer and requested allocation size outside of heap!"
+			);
+
 			let four_bytes = self.get_heap_4bytes(item)?;
 			self.heads[list_index] = FreeingBumpHeapAllocator::le_bytes_to_u32(four_bytes);
-			item + 8
+			ptr
 		} else {
 			// Nothing to be freed. Bump.
-			self.bump(item_size + 8) + 8
+			if self.bumper + PREFIX_SIZE + item_size > self.max_heap_size {
+				return Err(Error::AllocatorOutOfSpace);
+			}
+			self.bump(item_size + PREFIX_SIZE) + PREFIX_SIZE
```

Comment:
cleaner if `bump` returns a `Result`, then we don't access `self.bumper` from two abstraction levels.

```suggestion
			self.bump(item_size + PREFIX_SIZE)? + PREFIX_SIZE
```